app/src/test/java/com/github/se/cyrcle/model/image/ImageRepositoryCloudStorageTest.kt

```diff
@@ -0,0 +1,152 @@
+  @Mock private lateinit var mockContext: Context
+  @Mock private lateinit var mockContentResolver: ContentResolver
+
+  private lateinit var imageRepositoryCloudStorage: ImageRepositoryCloudStorage
+
+  @Before
+  fun setUp() {
+    MockitoAnnotations.openMocks(this)
+    `when`(mockFirebaseStorage.reference).thenReturn(mockStorageReference)
+    imageRepositoryCloudStorage = ImageRepositoryCloudStorage(mockAuth, mockFirebaseStorage)
+    Robolectric.setupActivity(Activity::class.java)
+  }
+
+  @Test
+  fun testGetUrl(): Unit {
+
+    val path = "test/path"
+    val expectedUrl = "https://example.com/test/path"
+    // set the task up
+    val taskCompletionSource = TaskCompletionSource<Uri>()
+    `when`(mockAuth.currentUser).thenReturn(mock())
+    `when`(mockStorageReference.child(path)).thenReturn(mockStorageReference)
+    // call the onSuccess lambda when .downloadUrl is called
+    `when`(mockStorageReference.downloadUrl).thenReturn(taskCompletionSource.task)
+
+    var actualUrl: String? = null
+    var onFailureCalled = false
+    var onSuccessCalled = false
+
+    imageRepositoryCloudStorage.getUrl(
+        path,
+        onSuccess = {
+          onSuccessCalled = true
+          actualUrl = it
+        },
+        onFailure = { onFailureCalled = true })
+    // resolve the task
+    taskCompletionSource.setResult(Uri.parse(expectedUrl))
+    shadowOf(Looper.getMainLooper()).idle()
+    assertTrue(onSuccessCalled)
+    assert(actualUrl == expectedUrl)
+    assertFalse(onFailureCalled)
+  }
+
+  @Test
+  fun testGetUrlWithNullUser() {
+    val path = "test/path"
+    `when`(mockAuth.currentUser).thenReturn(null)
+    var onFailureCalled = false
+    var onSuccessCalled = false
+    imageRepositoryCloudStorage.getUrl(
+        path, onSuccess = { onSuccessCalled = true }, onFailure = { onFailureCalled = true })
+    shadowOf(Looper.getMainLooper()).idle()
+    assertFalse(onSuccessCalled)
+    assertTrue(onFailureCalled)
+  }
+
+  @Test
+  fun testUploadImage() {
+    val mockTaskSnapshot = mock(UploadTask.TaskSnapshot::class.java)
+    val fileUri = "test/path"
+    val destinationPath = "test/path"
+    val expectedUrl = "https://example.com/test/path"
+    val bytearray = ByteArray(100)
+    val taskCompletionSource = TaskCompletionSource<Uri>()
+    val mockUploadTask = mock(UploadTask::class.java)
+    `when`(mockAuth.currentUser).thenReturn(mock())
+    `when`(mockContext.contentResolver).thenReturn(mockContentResolver)
+    `when`(mockContentResolver.openInputStream(Uri.parse(fileUri)))
+        .thenReturn(ByteArrayInputStream(bytearray))
+    `when`(mockStorageReference.child(destinationPath)).thenReturn(mockStorageReference)
+    `when`(mockStorageReference.putBytes(any())).thenReturn(mockUploadTask)
+    `when`(mockStorageReference.downloadUrl).thenReturn(taskCompletionSource.task)
+    `when`(mockUploadTask.addOnSuccessListener(any())).thenAnswer { invocation ->
```

Comment:
I admire your motivation to mock...

Reply:
I don't think I had the choice ... did I ?

Reply:
Not really :)